pkg/virt-launcher/virtwrap/network/podnic_test.go

```diff
@@ -0,0 +1,157 @@
+			address2 := &net.IPNet{IP: net.IPv4(169, 254, 0, 0)}
+			fakeAddr1 := netlink.Addr{IPNet: address1}
+			fakeAddr2 := netlink.Addr{IPNet: address2}
+			mockNetwork.EXPECT().ReadIPAddressesFromLink(primaryPodInterfaceName).Return(fakeAddr1.IP.String(), fakeAddr2.IP.String(), nil)
+			mockNetwork.EXPECT().IsIpv4Primary().Return(true, nil).Times(1)
+			mockPodNetworkConfigurator.EXPECT().DiscoverPodNetworkInterface(primaryPodInterfaceName).Times(1)
+			mockPodNetworkConfigurator.EXPECT().GenerateDHCPConfig().Return(&cache.DHCPConfig{}).Times(1)
+			mockPodNetworkConfigurator.EXPECT().GenerateDomainIfaceSpec().Times(1)
+			mockPodNetworkConfigurator.EXPECT().PreparePodNetworkInterface().Return(fmt.Errorf("podnic_test: forcing prepare networking failure")).Times(1)
+		})
+		It("should return a CriticalNetworkError at phase1", func() {
+
+			domain := NewDomainWithBridgeInterface()
+			vmi := newVMIBridgeInterface("testnamespace", "testVmName")
+
+			api.NewDefaulter(runtime.GOARCH).SetObjectDefaults_Domain(domain)
+
+			podnic, err := newPodNICWithMocks(vmi)
+			Expect(err).ToNot(HaveOccurred())
+			err = podnic.PlugPhase1()
+			Expect(err).To(HaveOccurred(), "SetupPhase1 should return an error")
+
+			_, ok := err.(*neterrors.CriticalNetworkError)
+			Expect(ok).To(BeTrue(), "SetupPhase1 should return an error of type CriticalNetworkError")
+		})
+	})
+	Context("when DHCP startup fails", func() {
+		BeforeEach(func() {
+			mockNetwork.EXPECT().StartDHCP(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(errors.New("podnic_test: forcing failure at DHCP start"))
+		})
+		It("phase2 should panic", func() {
+			testDhcpPanic := func() {
+				domain := NewDomainWithBridgeInterface()
+				vmi := newVMIBridgeInterface("testnamespace", "testVmName")
+				api.NewDefaulter(runtime.GOARCH).SetObjectDefaults_Domain(domain)
+				podnic, err := newPodNICWithMocks(vmi)
+				Expect(err).ToNot(HaveOccurred())
+				podnic.storeCachedDomainIface(domain.Spec.Devices.Interfaces[0])
+				podnic.dhcpConfigurator.ExportConfiguration(cache.DHCPConfig{Name: "eth0"})
+				Expect(podnic.PlugPhase2(domain)).To(Succeed())
+			}
+			Expect(testDhcpPanic).To(Panic())
+		})
+	})
+	Context("when setPodInterfaceCache is called with a configured nic", func() {
+		var (
+			podnic *podNIC
+		)
+		BeforeEach(func() {
+			vmi = newVMIMasqueradeInterface("testnamespace", "testVmName")
+			vmi.UID = types.UID("test-1234")
+			address1 := &net.IPNet{IP: net.IPv4(1, 2, 3, 4)}
+			address2 := &net.IPNet{IP: net.IPv4(169, 254, 0, 0)}
+			fakeAddr1 := netlink.Addr{IPNet: address1}
+			fakeAddr2 := netlink.Addr{IPNet: address2}
+
+			mockNetwork.EXPECT().ReadIPAddressesFromLink(primaryPodInterfaceName).Return(fakeAddr1.IP.String(), fakeAddr2.IP.String(), nil)
+			mockNetwork.EXPECT().IsIpv4Primary().Return(true, nil).Times(1)
+
+			var err error
+			podnic, err = newPodNICWithMocks(vmi)
+			Expect(err).ToNot(HaveOccurred())
+			err = podnic.setPodInterfaceCache()
```

Comment:
I know you just moved the already existing tests from podinterface_test, but I still believe it is a good opportunity to do/discuss how PodNIC should be unit tested.
newPodNICWithMocks is un-exported method. Do we want to unit test un-exported methods of podNIC? If so, why just this one?

Reply:
The `newPodNICWithMocks` is a helper function from the tests not from `podNIC` here we tests that the pod interface cache is written after reading the network configuration and the dependencie are mocked so the tests is all good.

Reply:
Sorry, I of-course meant `setPodInterfaceCache`.

Reply:
Now that we have a possitive test for Phase1 we can check the pod interface cache there, doing that.